src/DFAMin.hs

```diff
@@ -112,57 +131,47 @@ groupEquivStates DFA { dfa_states = statemap }
     accept_groups :: [EquivalenceClass]
     accept_groups = map IS.fromList (Map.elems accept_map)
 
-    init_p, init_q :: [EquivalenceClass]
-    init_p  -- Issue #71: each EquivalenceClass needs to be a non-empty set
-      | IS.null nonaccepting_states = accept_groups
-      | otherwise                   = nonaccepting_states : accept_groups
+    init_r, init_q :: [EquivalenceClass]
+    init_r  -- Issue #71: each EquivalenceClass needs to be a non-empty set
+      | IS.null nonaccepting_states = []
+      | otherwise                   = [nonaccepting_states]
     init_q = accept_groups
 
-    -- map token T to
-    --   a map from state S to the list of states that transition to
+    -- a map from token T to
+    --   a map from state S to the set of states that transition to
     --   S on token T
-    -- This is a cache of the information needed to compute x below
-    bigmap :: IntMap (IntMap [SNum])
-    bigmap = IM.fromListWith (IM.unionWith (++))
-                [ (i, IM.singleton to [from])
+    -- This is a cache of the information needed to compute xs below
+    bigmap :: IntMap (IntMap EquivalenceClass)
+    bigmap = IM.fromListWith (IM.unionWith IS.union)
+                [ (i, IM.singleton to (IS.singleton from))
                 | (from, state) <- Map.toList statemap,
                   (i,to) <- IM.toList (state_out state) ]
 
-    -- incoming I A = the set of states that transition to a state in
-    -- A on token I.
-    incoming :: Int -> IntSet -> IntSet
-    incoming i a = IS.fromList (concat ss)
-       where
-         map1 = IM.findWithDefault IM.empty i bigmap
-         ss = [ IM.findWithDefault [] s map1
-              | s <- IS.toList a ]
-
-    -- The outer loop: recurse on each set in Q
-    go ::  [EquivalenceClass] -> [EquivalenceClass] -> [EquivalenceClass]
-    go p [] = p
-    go p (a:q) = go1 0 p q
-     where
-       -- recurse on each token (0..255)
-       go1 256 p q = go p q
-       go1 i   p q = go1 (i+1) p' q'
+    -- The outer loop: recurse on each set in R and Q
+    go :: [EquivalenceClass] -> [EquivalenceClass] -> [EquivalenceClass]
+    go r [] = r
+    go r (a:q) = uncurry go $ List.foldl' go0 (a:r,q) xs
+      where
+        xs :: [EquivalenceClass]
+        xs = filter (not . IS.null)
+           . map (\m -> IS.unions [IM.findWithDefault IS.empty s m | s <- IS.toList a])
+           $ IM.elems bigmap
+
+        go0 (r,q) x = go1 r [] []
           where
-            (p',q') = go2 p [] q
-
-            x = incoming i a
+            go1 []    r' q' = (r', go2 q q')
+            go1 (y:r) r' q'
+              | IS.null y1 || IS.null y2 = go1 r (y:r') q'
+              | IS.size y1 <= IS.size y2 = go1 r (y2:r') (y1:q')
+              | otherwise                = go1 r (y1:r') (y2:q')
+              where
+                y1 = IS.intersection x y
+                y2 = IS.difference   y x
```

Comment:
Let's definitely make the
```
contradictions
  :: IntSet -- ^ initial equivalence class
  -> IntSet -- ^ primage single partition of the space. (N.B. Can also be converse, doesn't matter.)
  -> Maybe (IntSet, IntSet) -- N.B. both non-empty
```
helper for the implementation, whether or not we use it in the source code.

Reply:
I applied it https://github.com/simonmar/alex/pull/176/commits/f691e0e32c69dedea184bd6cbe96e0c497227fa9